### PR TITLE
Staged_root & partition in submit.yml.erb render

### PR DESCRIPTION
**Pull request by ericfranz**
**This is an alternative to https://github.com/OSC/ondemand/pull/915 and has not yet been tested**

Rendering context should be attribute name & values AND staged_root
So we convert the array of Attribute objects to a Hash of id => value pairs (Attribute's name is returned by method id)
and convert that to an OpenStruct, which gives the same "attribute_name as method" experience,
lets us add staged_root, and avoids conflicts with other predefined methods like partition

we have to define a method on the OpenStruct instance we create 'get_binding'

### CODE REVIEW

**Comment by ericfranz** [comment]
This is partly complete. Fixes https://github.com/OSC/ondemand/issues/605 as well - but is there another place where we want to call `SessionContext#attr_name=` or `SessionContext#attr_value`? The solution to the attribute named `partition` issue is to avoid using method_missing on a SessionContext Enumerable object to get and set the value of attributes for a SessionContext altogether! This fancy object tries to treat an array as a hash using method_missing - too much complexity for what its worth. Minimally I would

1.  remove method_missing and respond_to_missing? https://github.com/OSC/ondemand/blob/17ae54c99255389823d1850c208ec53ad02dad1a/apps/dashboard/app/models/batch_connect/session_context.rb#L50-L67; this should expose problems wherever we try to use batch connect apps
2. in those areas to fix, we may need a hash representation of SessionContext. add `SessionContext#to_h`, `#to_openstruct`, `#keys` methods, and use that there (and in the changes in this PR)
3. Eventually dumb down SessionContext so it doesn't even use `ActiveModel::Model` or `ActiveModel::Serializers::JSON`

Of course this is a bigger lift. If we just get staged_root in and save the rest of these changes for later that might be good too.

File: apps/dashboard/app/models/batch_connect/app.rb [hunk 1]
```diff
@@ -216,12 +216,28 @@ def build_session_context
     # @param session_context [SessionContext] object with attributes
     # @param fmt [String, nil] formatting used for attributes in submit hash
     # @return [Hash] hash of submission options
-    def submit_opts(session_context, fmt: nil)
+    def submit_opts(session_context, staged_root: staged_root, fmt: nil)
```
Comment (treydock) [inline apps/dashboard/app/models/batch_connect/app.rb:219]:
```suggestion
    def submit_opts(session_context, staged_root: nil, fmt: nil)
```
Get circular issues with doing `staged_root: staged_root`

**Comment by johrstrom** [comment]
Ya know, even I started to make a new PR based on this. I think the openstruct is a good idea, but say a user uses partition here. It works OK for the form -> submit.yml, but in the script.sh.erb they cannot use `context.partition` because _that_ context is the SessionContext which is Enumerable.

So I'm thinking we need to solve these issues separately.  #864 as a simpler PR now and defer #605 to something that does work well for both `template/*.erb` and `submit.yml.erb`.

**Comment by treydock** [comment]
This pull request solves #864, or do you mean a simpler approach than this pull request to get the staged_root into submit.yml.erb?  Sorry if I misunderstood.

**Comment by treydock** [comment]
@johrstrom Should this be closed in favor of #1159 ?

**Comment by johrstrom** [comment]
Yea this if fixed in #1159. I did push `staged-root-submit` which has session#to_openstruct API which we can refer to to fix #605 at some point.

**Pull request closed without merging**